openquake/hazardlib/gsim/mgmpe/generic_gmpe_avgsa.py

```diff
@@ -0,0 +1,198 @@
+# The Hazard Library
+# Copyright (C) 2012-2019 GEM Foundation
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Module :mod:`openquake.hazardlib.mgmp.generic_gmpe_avgsa` implements
+:class:`~openquake.hazardlib.mgmpe.GenericGmpeAvgSA`
+"""
+
+import copy
+import numpy as np
+from openquake.hazardlib.gsim.base import GMPE, registry
+from openquake.hazardlib import const
+from openquake.hazardlib.imt import SA
+from openquake.hazardlib.gsim.mgmpe import akkar_coeff_table as act
+
+class GenericGmpeAvgSA(GMPE):
+    """
+    Implements a modified GMPE class that can be used to compute average
+    ground motion over several spectral ordinates from an arbitrary GMPE.
+    The mean and standard deviation are computed according to:
+    Kohrangi M., Reddy Kotha S. and Bazzurro P., 2018, Ground-motion models
+    for average spectral acceleration in a period range: direct and indirect
+    methods, Bull. Earthquake. Eng., 16, pp. 45–65.
+    Note that only the Total Standard Deviation is supported.
+
+    :param string gmpe_name:
+        The name of a GMPE class used for the calculation.
+
+    :param list avg_periods:
+        List of averaging periods (must be a subset of the periods allowed
+        in the selected GMPE)
+
+    :param string corr_func:
+        Handle of the function to compute correlation coefficients between
+        different spectral acceleration ordinates. Valid options are:
+        'baker_jayaram', 'akkar', 'none'. Default is none.
+    """
+
+    # Parameters
+    REQUIRES_SITES_PARAMETERS = set()
+    REQUIRES_DISTANCES = set()
+    REQUIRES_RUPTURE_PARAMETERS = set()
+    DEFINED_FOR_INTENSITY_MEASURE_COMPONENT = ''
+    DEFINED_FOR_INTENSITY_MEASURE_TYPES = set([SA])
+    DEFINED_FOR_STANDARD_DEVIATION_TYPES = set([const.StdDev.TOTAL])
+    DEFINED_FOR_TECTONIC_REGION_TYPE = ''
+    DEFINED_FOR_REFERENCE_VELOCITY = None
+
+    def __init__(self, gmpe_name, avg_periods, corr_func='none'):
+        super().__init__(gmpe_name=gmpe_name)
+        self.gmpe = registry[gmpe_name]()
+        self.set_parameters()
+        self.avg_periods = [float(t) for t in avg_periods.split(',')]
+        self.tnum = len(self.avg_periods)
+
+        correlation_function_handles = {
```

Comment:
Rather than this, I think it would have been more flexible to create an abstract class for a correlation model.

Reply:
That might be a good idea. @micheles , what's your opinion on that? Could you help me in case?

Reply:
An abstract class for the correlation model has been implemented. The previous correlation functions are now derived class.